keps/sig-apps/0008-sidecarcontainer.md

```diff
@@ -0,0 +1,116 @@
+---
+kep-number: 0008
+title: Sidecar Containers
+authors:
+  - "@joseph-irving"
+owning-sig: sig-apps
+participating-sigs:
+  - sig-apps
+  - sig-node
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2018-05-14
+last-updated: 2018-05-14
+status: provisional
+---
+
+# Sidecar Containers
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Alternatives](#alternatives-optional)
+
+## Summary
+
+To solve the issue of sidecars not being well supported we can create a new type of container called a sidecar which are handled like normal containers but with the caveat that running sidecar containers will always be terminated when all non-sidecar containers have terminated. This will solve the problem of Jobs not completing/failing when a sidecar container never exits due to it having no awareness of the 'primary' container.
+
+## Motivation
+
+SideCar containers have always been used in some ways but just not formally identified as such, with the introduction of Jobs and Cronjobs it became clear that they're not handled very well. If you have a Job with two containers one of which is actually doing the main processing of the job and the other is just facilitating it, you encounter a problem when the main process finishes; your sidecar container will carry on running so the job will never finish.
+
+The only way around this problem is to manage the sidecar container's lifecycle manually and arrange for it to exit when the main container exits. This is typically achieved by building an ad-hoc signalling mechanism to communicate completion status between containers. Common implementations use a shared scratch volume mounted into all pods, where lifecycle status can be communicated by creating and watching for the presence of files. This pattern has several disadvantages:
+
+* Repetitive lifecycle logic must be rewritten in each instance a sidecar is deployed.
+* Third-party containers typically require a wrapper to add this behaviour, normally provided via an entrypoint wrapper script implemented in the k8s container spec. This adds undesirable overhead and introduces repetition between the k8s and upstream container image specs.
+* The wrapping typically requires the presence of a shell in the container image, so this pattern does not work for minimal containers which ship without a toolchain.
+
+### Goals
+
+Solve issue [25908](https://github.com/kubernetes/kubernetes/issues/25908) so that sidecars can be easily handled without needing modifications to the container itself
```

Comment:
This proposal can potentially solve issue [52345](https://github.com/kubernetes/kubernetes/issues/52345) as well.

Reply:
Are there other longstanding pod issues that this proposal can resolve?  Making a case that a larger class of use cases is resolved by a simpler proposal gives this approach some weight.

Reply:
@smarterclayton so far all the issues this solves seem to just be variations of the original issue [25908](https://github.com/kubernetes/kubernetes/issues/25908), if you think it would be useful to list these we could do that.

If anyone has any other issues that will be solved by this proposal please do put them in a comment

Reply:
I would vote for including the use-case from [52345](https://github.com/kubernetes/kubernetes/issues/52345). That would make this proposal really about the `pod` and not only about `jobs`. But that would mean taking the handling of restart policy `Always` into the scope of this proposal. It seems currently to be kept out of scope.

Reply:
So as I understand it you want a way to cause your entire pod to restart when a primary container restarts so that init containers re-run. This is almost the same as starting a brand new pod. So you could do this with the current proposal by setting restart policy to `Never`, then when your primary container exits for whatever reason, the whole pod will be terminated and a new pod will start. However, this doesn't work in the case of having multiple primary containers as you need all of them to die before the pod is terminated.

@derekwaynecarr suggested having a field of `terminationFatalToPod: true` which sounds like it could solve your problem but I don't think it solves the sidecar problem. So maybe they should be separate proposals.

Reply:
Unfortunately, `restartPolicy` `Never` is not allowed with `ReplicaSets`, `Deployments` and `StatefulSets`.